src/mds/CInode.cc

```diff
@@ -697,48 +697,36 @@ frag_t InodeStoreBase::pick_dirfrag(std::string_view dn)
   return dirfragtree[h];
 }
 
-std::pair<bool, std::vector<CDir*>> CInode::get_dirfrags_under(frag_t fg)
+std::vector<CDir*> CInode::get_dirfrags_under(frag_t fg, bool* all )
 {
-  std::pair<bool, std::vector<CDir*>> result;
-  auto& all = result.first;
-  auto& dirs = result.second;
+  std::vector<CDir*> dirs;
 
-  {
-    frag_vec_t leaves;
-    dirfragtree.get_leaves_under(fg, leaves);
-    for (const auto &leaf : leaves) {
-      if (auto it = dirfrags.find(leaf); it != dirfrags.end()) {
-        dirs.push_back(it->second);
-      } else {
-        all = false;
-      }
-    }
-  }
-
-  if (all)
-    return result;
-
-  fragtree_t tmpdft;
-  tmpdft.force_to_leaf(g_ceph_context, fg);
-  for (auto &p : dirfrags) {
-    tmpdft.force_to_leaf(g_ceph_context, p.first);
-    if (fg.contains(p.first) && !dirfragtree.is_leaf(p.first))
-      dirs.push_back(p.second);
+  if (auto it = dirfrags.find(fg); it != dirfrags.end()){
+    if(all)
+      *all = true;
+    dirs.push_back(it->second);
+    return dirs;
   }
-
-  all = true;
-  {
-    frag_vec_t leaves;
-    tmpdft.get_leaves_under(fg, leaves);
-    for (const auto& leaf : leaves) {
-      if (!dirfrags.count(leaf)) {
-        all = false;
-        break;
+  
+  int total = 0;
+  for(auto &p : dirfrags) {
+    if (fg.bits() >= p.first.bits()) {
+      if (p.first.contains(fg)) {
+	if (all)
+	  *all = true;
```

Comment:
Forgot to include `p` in results?

Reply:
In this case, the original output of this function does not contain this partition.

Reply:
> Forgot to include `p` in results?

In this case, the original output of this function does not contain this partition, I think